c++/src/io/Cache.hh

```diff
@@ -0,0 +1,230 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <orc/MemoryPool.hh>
+#include <orc/OrcFile.hh>
+
+#include <algorithm>
+#include <cassert>
+#include <cstdint>
+#include <future>
+#include <utility>
+#include <vector>
+
+namespace orc {
+  class InputStream;
+
+  struct CacheOptions {
+    /// The maximum distance in bytes between two consecutive
+    /// ranges; beyond this value, ranges are not combined
+    uint64_t holeSizeLimit = 8192;
+
+    /// The maximum size in bytes of a combined range; if
+    /// combining two consecutive ranges would produce a range of a
+    /// size greater than this, they are not combined
+    uint64_t rangeSizeLimit = 32 * 1024 * 1024;
+  };
+
+  struct ReadRange {
+    uint64_t offset;
+    uint64_t length;
+
+    ReadRange() = default;
+    ReadRange(uint64_t offset, uint64_t length) : offset(offset), length(length) {}
+
+    friend bool operator==(const ReadRange& left, const ReadRange& right) {
+      return (left.offset == right.offset && left.length == right.length);
+    }
+    friend bool operator!=(const ReadRange& left, const ReadRange& right) {
+      return !(left == right);
+    }
+
+    bool contains(const ReadRange& other) const {
+      return (offset <= other.offset && offset + length >= other.offset + other.length);
+    }
+  };
+
+  struct ReadRangeCombiner {
+    std::vector<ReadRange> coalesce(std::vector<ReadRange> ranges) const {
+      if (ranges.empty()) {
+        return ranges;
+      }
+
+      // Remove zero-sized ranges
+      auto end = std::remove_if(ranges.begin(), ranges.end(),
+                                [](const ReadRange& range) { return range.length == 0; });
+      // Sort in position order
+      std::sort(ranges.begin(), end,
+                [](const ReadRange& a, const ReadRange& b) { return a.offset < b.offset; });
+
+      // Remove ranges that overlap 100%
+      end = std::unique(ranges.begin(), end, [](const ReadRange& left, const ReadRange& right) {
```

Comment:
The binary predicate does not establish an equivalence relation, the behavior is undefined. https://en.cppreference.com/w/cpp/algorithm/unique

Reply:
fixed.